clients/python/tensorzero/types.py

```diff
@@ -372,14 +372,61 @@ class VariantExtraBody(TypedDict):
     delete: NotRequired[bool]
 
 
+# DEPRECATED: Use `ModelProviderExtraBody` instead.
 class ProviderExtraBody(TypedDict):
     model_provider_name: str
     pointer: str
     value: NotRequired[Any]
     delete: NotRequired[bool]
 
 
-ExtraBody = Union[VariantExtraBody, ProviderExtraBody]
+class ModelProviderExtraBody(TypedDict):
+    model_name: str
+    provider_name: str
+    pointer: str
+    value: NotRequired[Any]
+    delete: NotRequired[bool]
+
+
+class AlwaysExtraBody(TypedDict):
+    pointer: str
+    value: NotRequired[Any]
+    delete: NotRequired[bool]
+
+
+ExtraBody = Union[VariantExtraBody, ProviderExtraBody, ModelProviderExtraBody, AlwaysExtraBody]
+
+
+class VariantExtraHeader(TypedDict):
+    variant_name: str
+    name: str
+    value: NotRequired[str]
+    delete: NotRequired[bool]
+
+
+# DEPRECATED: Use `ModelProviderExtraHeader` instead.
+class ProviderExtraHeader(TypedDict):
+    model_provider_name: str
+    name: str
+    value: NotRequired[str]
+    delete: NotRequired[bool]
+
+
+class ModelProviderExtraHeader(TypedDict):
+    model_name: str
+    provider_name: str
+    name: str
+    value: NotRequired[str]
+    delete: NotRequired[bool]
+
+
+class AlwaysExtraHeader(TypedDict):
+    name: str
+    value: NotRequired[str]
+    delete: NotRequired[bool]
+
+
+ExtraHeader = Union[VariantExtraHeader, ProviderExtraHeader, ModelProviderExtraHeader, AlwaysExtraHeader]
```

Comment:
why are we handwriting these types? can we generate them?

Reply:
https://github.com/tensorzero/tensorzero/issues/4656